From 8fa8cdaf858be266de4bd1c6a5d1c609a83f5b75 Mon Sep 17 00:00:00 2001
From: Arun Raghavan <arunr@valvesoftware.com>
Date: Tue, 19 May 2026 11:52:38 -0700
Subject: [PATCH] Add a custom platform profile mechanism

Allows specifying cpufreq ranges manually, and can be extended to offer
other mechanisms in the future.
---
 data/devices/steam-frame.toml       |  31 +++++++++
 steamos-manager/src/hardware.rs     |  16 ++++-
 steamos-manager/src/manager/root.rs |   4 +-
 steamos-manager/src/manager/user.rs |   3 +-
 steamos-manager/src/power.rs        | 103 ++++++++++++++++++++++++++--
 5 files changed, 146 insertions(+), 11 deletions(-)

diff --git a/data/devices/steam-frame.toml b/data/devices/steam-frame.toml
index af265fa..d9abc52 100644
--- a/data/devices/steam-frame.toml
+++ b/data/devices/steam-frame.toml
@@ -9,3 +9,34 @@ oui = "e0-31-9e"
 [gpu_performance]
 driver = "devfreq"
 sysfs_path = "/sys/devices/platform/soc@0/3d00000.gpu/devfreq/3d00000.gpu"
+
+[performance_profile]
+platform_profile_driver = "custom"
+suggested_default = "Performance"
+
+[[performance_profile.custom_profile]]
+name = "Performance"
+cpufreq = [
+  { policy = 0, max = 2265600 },
+  { policy = 2, max = 3148800 },
+  { policy = 5, max = 2956800 },
+  { policy = 7, max = 3052800 },
+]
+
+[[performance_profile.custom_profile]]
+name = "Balanced"
+cpufreq = [
+  { policy = 0, max = 2265600 },
+  { policy = 2, max = 2515200 },
+  { policy = 5, max = 2515200 },
+  { policy = 7, max = 2553600 },
+]
+
+[[performance_profile.custom_profile]]
+name = "Powersave"
+cpufreq = [
+  { policy = 0, max = 2035200 },
+  { policy = 2, max = 2035200 },
+  { policy = 5, max = 2035200 },
+  { policy = 7, max = 2035200 },
+]
diff --git a/steamos-manager/src/hardware.rs b/steamos-manager/src/hardware.rs
index 80fcb99..e8b0412 100644
--- a/steamos-manager/src/hardware.rs
+++ b/steamos-manager/src/hardware.rs
@@ -184,9 +184,23 @@ pub(crate) struct InputPlumberConfig {
 #[derive(Clone, Deserialize, Debug)]
 pub(crate) struct PerformanceProfileConfig {
     pub suggested_default: String,
-    pub platform_profile_name: String,
+    pub platform_profile_name: Option<String>,
     #[serde(default)]
     pub platform_profile_driver: PlatformProfileDriverType,
+    pub custom_profile: Option<Vec<CustomPerformanceProfile>>,
+}
+
+#[derive(Clone, Deserialize, Debug)]
+pub(crate) struct CustomPerformanceProfile {
+    pub name: String,
+    pub cpufreq: Vec<CpufreqRange>,
+}
+
+#[derive(Clone, Deserialize, Debug)]
+pub(crate) struct CpufreqRange {
+    pub policy: u32,
+    pub min: Option<u32>,
+    pub max: Option<u32>,
 }
 
 #[derive(Clone, Deserialize, Debug)]
diff --git a/steamos-manager/src/manager/root.rs b/steamos-manager/src/manager/root.rs
index f27a3aa..c79171b 100644
--- a/steamos-manager/src/manager/root.rs
+++ b/steamos-manager/src/manager/root.rs
@@ -720,8 +720,8 @@ impl SteamOSManager {
         Ok(())
     }
 
-    async fn set_performance_profile(&self, profile: &str) -> fdo::Result<()> {
-        let Some(ref driver) = self.platform_profile else {
+    async fn set_performance_profile(&mut self, profile: &str) -> fdo::Result<()> {
+        let Some(ref mut driver) = self.platform_profile else {
             return Err(fdo::Error::Failed(String::from(
                 "Platform profile settings not configured",
             )));
diff --git a/steamos-manager/src/manager/user.rs b/steamos-manager/src/manager/user.rs
index 59c49dc..5d8a9f1 100644
--- a/steamos-manager/src/manager/user.rs
+++ b/steamos-manager/src/manager/user.rs
@@ -2238,7 +2238,8 @@ mod test {
             }),
             performance_profile: Some(PerformanceProfileConfig {
                 platform_profile_driver: PlatformProfileDriverType::Acpi,
-                platform_profile_name: String::from("power-driver"),
+                platform_profile_name: Some(String::from("power-driver")),
+                custom_profile: None,
                 suggested_default: String::from("balanced"),
             }),
             inputplumber: None,
diff --git a/steamos-manager/src/power.rs b/steamos-manager/src/power.rs
index 83e5251..f2ec3a0 100644
--- a/steamos-manager/src/power.rs
+++ b/steamos-manager/src/power.rs
@@ -31,7 +31,7 @@ use zbus::{Connection, ObjectServer, fdo};
 
 use crate::error::{to_zbus_error, to_zbus_fdo_error};
 use crate::gpu::AMDGPU_HWMON_NAME;
-use crate::hardware::{FanControlState, device_config};
+use crate::hardware::{CpufreqRange, CustomPerformanceProfile, FanControlState, device_config};
 use crate::manager::MANAGER_PATH;
 use crate::manager::root::RootManagerProxy;
 use crate::manager::user::TdpLimit1;
@@ -57,6 +57,9 @@ const CPU_POLICY_NAME: &str = "policy";
 const CPU_SCALING_GOVERNOR_SUFFIX: &str = "scaling_governor";
 const CPU_SCALING_AVAILABLE_GOVERNORS_SUFFIX: &str = "scaling_available_governors";
 
+const CPU_SCALING_MAX_FREQ_SUFFIX: &str = "scaling_max_freq";
+const CPU_SCALING_MIN_FREQ_SUFFIX: &str = "scaling_min_freq";
+
 const LAVD_PATH: &str = "/usr/bin/scx_lavd";
 
 const PLATFORM_PROFILE_PREFIX: &str = "/sys/class/platform-profile";
@@ -130,11 +133,27 @@ pub(crate) async fn platform_profile_driver() -> Result<Box<dyn PlatformProfileD
         .and_then(|device_config| device_config.performance_profile.as_ref())
         .ok_or(anyhow!("No platform profile driver configured"))?;
 
-    Ok(match &config.platform_profile_driver {
+    match &config.platform_profile_driver {
         PlatformProfileDriverType::Acpi => {
-            Box::new(AcpiPlatformProfileDriver::new(&config.platform_profile_name).await?)
+            if let Some(ref platform_profile_name) = config.platform_profile_name {
+                Ok(Box::new(
+                    AcpiPlatformProfileDriver::new(platform_profile_name).await?,
+                ))
+            } else {
+                Err(anyhow!("No platform profile name configured"))
+            }
         }
-    })
+        PlatformProfileDriverType::Custom => {
+            if let Some(ref profile) = config.custom_profile {
+                Ok(Box::new(
+                    CustomPlatformProfileDriver::new(profile, Some(&config.suggested_default))
+                        .await?,
+                ))
+            } else {
+                Err(anyhow!("No custom profiles configured"))
+            }
+        }
+    }
 }
 
 #[derive(Deserialize, Display, EnumString, VariantNames, PartialEq, Debug, Clone)]
@@ -258,6 +277,32 @@ async fn read_cpu_sysfs_contents<S: AsRef<Path>>(suffix: S) -> Result<String> {
         .map_err(|message| anyhow!("Error opening sysfs file for reading {message}"))
 }
 
+async fn write_cpu_freq_sysfs_contents(range: &CpufreqRange) -> Result<()> {
+    let base = path(CPU_PREFIX)
+        .join(CPUFREQ_PREFIX)
+        .join(format!("policy{}", range.policy));
+
+    if let Some(max) = range.max {
+        write_synced(
+            base.join(CPU_SCALING_MAX_FREQ_SUFFIX),
+            max.to_string().as_bytes(),
+        )
+        .await
+        .inspect_err(|message| error!("Error writing to sysfs file: {message}"))?;
+    }
+
+    if let Some(min) = range.min {
+        write_synced(
+            base.join(CPU_SCALING_MIN_FREQ_SUFFIX),
+            min.to_string().as_bytes(),
+        )
+        .await
+        .inspect_err(|message| error!("Error writing to sysfs file: {message}"))?;
+    }
+
+    Ok(())
+}
+
 async fn write_cpu_governor_sysfs_contents(contents: String) -> Result<()> {
     // Iterate over all policyX paths
     let mut dir = read_dir(path(CPU_PREFIX).join(CPUFREQ_PREFIX)).await?;
@@ -735,6 +780,7 @@ pub(crate) async fn set_max_charge_level(limit: i32) -> Result<oneshot::Receiver
 pub(crate) enum PlatformProfileDriverType {
     #[default] // historical default
     Acpi,
+    Custom,
 }
 
 #[derive(Debug)]
@@ -742,6 +788,12 @@ pub(crate) struct AcpiPlatformProfileDriver {
     path: PathBuf,
 }
 
+#[derive(Debug)]
+pub(crate) struct CustomPlatformProfileDriver {
+    profiles: Vec<CustomPerformanceProfile>,
+    current_profile: Option<String>,
+}
+
 #[async_trait]
 pub(crate) trait PlatformProfileDriver: Send + Sync {
     // This can be used from the user and root managers
@@ -749,7 +801,7 @@ pub(crate) trait PlatformProfileDriver: Send + Sync {
     // This can be used from the user and root managers
     async fn get_platform_profile(&self) -> Result<String>;
     // This can only be used from the root manager
-    async fn set_platform_profile(&self, profile: &str) -> Result<()>;
+    async fn set_platform_profile(&mut self, profile: &str) -> Result<()>;
 }
 
 impl AcpiPlatformProfileDriver {
@@ -779,13 +831,49 @@ impl PlatformProfileDriver for AcpiPlatformProfileDriver {
             .to_string())
     }
 
-    async fn set_platform_profile(&self, profile: &str) -> Result<()> {
+    async fn set_platform_profile(&mut self, profile: &str) -> Result<()> {
         fs::write(self.path.join("profile"), profile.as_bytes())
             .await
             .map_err(|message| anyhow!("Error writing to sysfs: {message}"))
     }
 }
 
+impl CustomPlatformProfileDriver {
+    async fn new(profiles: &[CustomPerformanceProfile], current: Option<&String>) -> Result<Self> {
+        Ok(Self {
+            profiles: profiles.to_owned(),
+            current_profile: current.cloned(),
+        })
+    }
+}
+
+#[async_trait]
+impl PlatformProfileDriver for CustomPlatformProfileDriver {
+    async fn get_available_platform_profiles(&self) -> Result<Vec<String>> {
+        Ok(self.profiles.iter().map(|p| p.name.clone()).collect())
+    }
+
+    async fn get_platform_profile(&self) -> Result<String> {
+        if let Some(name) = &self.current_profile {
+            Ok(name.clone())
+        } else {
+            Err(anyhow!("No profile is currently set"))
+        }
+    }
+
+    async fn set_platform_profile(&mut self, profile: &str) -> Result<()> {
+        if let Some(profile) = self.profiles.iter().find(|p| p.name == profile) {
+            for cpufreq in &profile.cpufreq {
+                write_cpu_freq_sysfs_contents(cpufreq).await?
+            }
+            self.current_profile = Some(profile.name.clone());
+            Ok(())
+        } else {
+            Err(anyhow!("Unknown custom profile '{profile}'"))
+        }
+    }
+}
+
 pub(crate) async fn register_tdp_limit1(
     ctx: &mut Option<UnboundedSender<TdpManagerCommand>>,
     proxy: TdpLimit1Proxy<'_>,
@@ -1941,7 +2029,8 @@ pub(crate) mod test {
         let config = DeviceConfig {
             performance_profile: Some(PerformanceProfileConfig {
                 platform_profile_driver: PlatformProfileDriverType::Acpi,
-                platform_profile_name: String::from("platform-profile0"),
+                platform_profile_name: Some(String::from("platform-profile0")),
+                custom_profile: None,
                 suggested_default: String::from("custom"),
             }),
             tdp_limit: Some(TdpLimitConfig {
